meta/rpc.go

```diff
@@ -0,0 +1,445 @@
+package meta
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/influxdb/influxdb/meta/internal"
+)
+
+// Max size of a message before we treat the size as invalid
+const (
+	MaxMessageSize    = 1024 * 1024 * 1024
+	leaderDialTimeout = 10 * time.Second
+)
+
+// RPC handles request/response style messaging between cluster nodes
+type RPC struct {
+	Logger         *log.Logger
+	tracingEnabled bool
+
+	store interface {
+		cachedData() *Data
+		IsLeader() bool
+		Leader() string
+		Peers() []string
+		AddPeer(host string) error
+		CreateNode(host string) (*NodeInfo, error)
+		NodeByHost(host string) (*NodeInfo, error)
+		WaitForDataChanged() error
+	}
+}
+
+type JoinResult struct {
+	RaftEnabled bool
+	RaftNodes   []string
+	NodeID      uint64
+}
+
+type Reply interface {
+	GetHeader() *internal.ResponseHeader
+}
+
+// proxyLeader proxies the connection to the current raft leader
+func (r *RPC) proxyLeader(conn *net.TCPConn) {
+	if r.store.Leader() == "" {
+		r.sendError(conn, "no leader")
+		return
+	}
+
+	leaderConn, err := net.DialTimeout("tcp", r.store.Leader(), leaderDialTimeout)
+	if err != nil {
+		r.sendError(conn, fmt.Sprintf("dial leader: %v", err))
+		return
+	}
+	defer leaderConn.Close()
+
+	leaderConn.Write([]byte{MuxRPCHeader})
+	if err := proxy(leaderConn.(*net.TCPConn), conn); err != nil {
+		r.sendError(conn, fmt.Sprintf("leader proxy error: %v", err))
+	}
+}
```

Comment:
Should the RPC simply tell the caller to retry against the leader instead? That way we don't have the overhead of proxying.